[MLIR][LLVM] Add vararg support in LLVM::CallOp

ivanradanov opened the pull request:
Adds an attribute that specifies the callee type of a LLVMCallOp. This is needed to distinguish between vararg and non-vararg calls as currently the LLVM Dialect always translates CallOp's to non-vararg calls.

The new syntax I opted for is this:

```
%2 = llvm.call @printf vararg !llvm.func<i32 (ptr<4>, ...)> (%1, %arg0) : (!llvm.ptr<4>, i32) -> i32
```

for vararg calls, and for non-vararg calls, the function type is omitted and the type will be inferred from the operation operand and result types, for example:

```
%2 = llvm.call @foo(%1, %arg0) : (!llvm.ptr<4>, i32) -> i32
```

I think it makes sense to keep the type at the end after the `:` which describes the type as a MLIR operation and the function type separate.

If this syntax looks good I will update the tests to reflect this change.

ivanradanov committed:
Build LLVM::CallOp with callee type info

ivanradanov committed:
Fix Conversion of CallOp func type

ivanradanov committed:
Print vararg info

ivanradanov commented:
@ftynse @wsmoses

gysit reviewed (commented):
Thanks for working on this.

I think personally I would put the vararg type after the arguments:
```
%2 = llvm.call @printf(%1, %arg0) vararg(!llvm.func<i32 (ptr<4>, ...)>) : (!llvm.ptr<4>, i32) -> i32
```
and wrap it into parentheses to emphasis stuff that is connected. The main advantage of this syntax is that the callee symbol and its arguments remain closely connected. What do you think? 

Overall this is mostly a matter of taste though.

ftynse reviewed (changes requested):
Repeating some types feels a bit tautological, but I don't have a better suggestion for the syntax at this point. Requesting changes due to a memory problem in the code.

ivanradanov committed:
Fixes

ivanradanov committed:
Also add callee type to invoke op

ivanradanov committed:
Add builder with func type to InvokeOp

ivanradanov committed:
Fix importing of LLVM Call and Invoke Instruction

ivanradanov committed:
Add a builder for indirect calls

ivanradanov committed:
Remove now supported case from failing test

ivanradanov committed:
Add test for importing vararg indirect call

ivanradanov commented:
Sorry I think I needed to rebase and force push because the code_formatter runner was getting confused?

Anyways, @gysit I liked your suggestion for the syntax so I have implemented that - also I did the same changes to InvokeOp and fixed the llvm<->mlir module translation and import and added a test.

Dinistro commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -1412,12 +1542,26 @@ ParseResult InvokeOp::parse(OpAsmParser &parser, OperationState &result) {
+  if (isVarArg) {
+    if (parser.parseLParen().failed() ||
+        !parser
+             .parseOptionalAttribute(calleeType, "callee_type",
```
NIT: As above. Also here, we should consider reusing the logic eventually (doesn't need to be this revision).

Dinistro commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -1412,12 +1542,26 @@ ParseResult InvokeOp::parse(OpAsmParser &parser, OperationState &result) {
   // Parse the trailing type list and resolve the function operands.
-  if (parseCallTypeAndResolveOperands(parser, result, isDirect, operands))
+  if (parseCallTypeAndResolveOperands(parser, result, isDirect, operands,
+                                      /*isVarArg*/ false))
```
```suggestion
                                      /*isVarArg=*/false))
```

Dinistro commented on mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp:
```diff
@@ -211,9 +211,15 @@ convertOperationImpl(Operation &opInst, llvm::IRBuilderBase &builder,
-                                                      callOp.getArgOperands()),
-                                operandsRef.front(), operandsRef.drop_front());
+      llvm::FunctionType *calleeType;
+      if (callOp.getCalleeType().has_value())
```
NIT: `has_value()` is not necessary
```suggestion
      if (callOp.getCalleeType())
```

Dinistro commented on mlir/lib/Target/LLVMIR/ModuleImport.cpp:
```diff
@@ -1411,20 +1418,25 @@ LogicalResult ModuleImport::convertInstruction(llvm::Instruction *inst) {
       invokeOp = builder.create<InvokeOp>(
-          loc, types, operands, directNormalDest, ValueRange(),
-          lookupBlock(invokeInst->getUnwindDest()), unwindArgs);
+          loc, funcTy, /*callee*/ nullptr, operands, directNormalDest,
```
```suggestion
          loc, funcTy, /*callee=*/nullptr, operands, directNormalDest,
```

Dinistro reviewed (commented):
Thanks for pushing on this. I dropped a bunch of nit comments and questions.
I think the design is reasonable, but I'm not yet sure if your code is fully correct.

gysit reviewed (commented):
I like the changes you did to the printer and parser :).

Can you add some more tests? I think it would make sense to have a roundtrip test to check the printing and parsing in mlir/test/Dialect/LLVMIR/func.mlir. Additionally, there should be a test that the translation to LLVM IR works as well.

ftynse commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -1148,14 +1201,8 @@ LogicalResult CallOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
-    return emitOpError()
-           << "indirect calls to variadic functions are not supported";
+  if (funcType.isVarArg() && !getCalleeType().has_value())
+    return emitOpError() << "Missing callee type attribute for vararg call";
```
Also nit: don't start error messages with a capital.

ftynse commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -1301,14 +1378,51 @@ ParseResult CallOp::parse(OpAsmParser &parser, OperationState &result) {
 
+  bool isVarArg = parser.parseOptionalKeyword("vararg").succeeded();
+  if (isVarArg) {
+    if (parser.parseLParen().failed() ||
```
IMO, explicit `.failed()` makes this more readable than the "true-means-error" convention that was originally borrowed from LLVM and is only used in MLIR parsers.

ivanradanov committed:
Update mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp

Co-authored-by: Christian Ulmann <christianulmann@gmail.com>

ivanradanov commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -1262,6 +1325,18 @@ static ParseResult parseCallTypeAndResolveOperands(
   if (funcType.getNumResults() != 0)
     result.addTypes(funcType.getResults());
 
+  if (!isVarArg) {
```
It is required if the callee is vararg, however, it can still be present for a non-vararg call, and we do not need (or want) to print it if the callee is non-vararg as the type is the same as the mlir op type printed after the `:`.

ivanradanov replied:
Oh you are indeed correct, we do not need to set that.

ivanradanov committed:
Update mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp

Co-authored-by: Christian Ulmann <christianulmann@gmail.com>

ivanradanov committed:
Comments

ivanradanov committed:
fix grammar

ivanradanov committed:
Fix wrong parseAttrib call

ivanradanov commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -1018,6 +1018,27 @@ static void printStoreType(OpAsmPrinter &printer, Operation *op,
+    resultType = LLVMVoidType::get(builder.getContext());
+  else
+    resultType = results.front();
+  return LLVMFunctionType::get(resultType, llvm::to_vector(args.getTypes()),
```
```
tools/mlir/include/mlir/Dialect/LLVMIR/LLVMTypes.h.inc:108:27: note: candidate function not viable: no known conversion from 'mlir::ValueRange::type_range' (aka 'ValueTypeRange<mlir::ValueRange>') to 'ArrayRef<mlir::Type>' for 2nd argument
  static LLVMFunctionType get(Type result, ArrayRef<Type> arguments, bool isVarArg = false);
```
I am getting this hence I added the `to_vector`

ftynse replied:
Yeah, it can be converted to `TypeRange` but not to `ArrayRef<Type>`. The caller needs to be updated to take `TypeRange` instead, but it's out of scope for this PR.

ivanradanov committed:
Add back to_vector

ivanradanov commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -1018,6 +1018,27 @@ static void printStoreType(OpAsmPrinter &printer, Operation *op,
 //===----------------------------------------------------------------------===//
 
+/// Get the MLIR Op-like result types of a LLVMFunctionType
+static SmallVector<Type, 1> getCallOpResults(LLVMFunctionType calleeType) {
```
I agree but I opted not to change the existing builders so as not to break code using them and adding even more builders started looking like way too many.

ivanradanov committed:
Do not start error messages with a capital

ivanradanov committed:
Nit

ivanradanov committed:
Wrong arg

ivanradanov committed:
Update mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp

Co-authored-by: Christian Ulmann <christianulmann@gmail.com>

ivanradanov committed:
Update mlir/lib/Target/LLVMIR/ModuleImport.cpp

Co-authored-by: Christian Ulmann <christianulmann@gmail.com>

ivanradanov committed:
invoke grammar

ivanradanov committed:
Fix LLVM::CallOp description

ivanradanov committed:
Test CallOp verifier

ivanradanov committed:
Add to-llvm test

ivanradanov commented:
Addressed most concerns and added mlir->llvm test.

I still need to add a roundtrip test

ivanradanov committed:
Fix printing

ivanradanov committed:
Fix tests

ivanradanov committed:
Roundtrip test

ivanradanov commented:
Done with suggested changes and edited the op description as well - please let me know if there are any other concerns

ivanradanov committed:
Fix toy example lowering to llvm

gysit commented on mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:
```diff
@@ -615,10 +611,17 @@ def LLVM_CallOp : LLVM_MemAccessOpBase<"call",
     // Indirect call with an argument and without a result.
     llvm.call %1(%0) : !llvm.ptr, (f32) -> ()
+
+    // Direct varaidic call
```
```suggestion
    // Direct variadic call.
```

ivanradanov committed:
Fix formatting

gysit reviewed (commented):
Thanks for the updates.

Can you test the invoke as well? I know it is almost the same logic but it may still be nice to have test coverage.

ivanradanov committed:
Apply suggestions from code review

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

ivanradanov committed:
Rename getCallOpResults -> getCallOpResultTypes

ivanradanov committed:
Do not set the calle type attrib if not variadic

ivanradanov committed:
Fix tests

ivanradanov committed:
Invoke tests

ivanradanov committed:
Invoke translation test

ivanradanov committed:
Add import invoke test

ivanradanov commented:
Thank you for the comments, I have added invoke tests as well.

gysit commented on mlir/test/Dialect/LLVMIR/roundtrip.mlir:
```diff
@@ -427,8 +437,21 @@ llvm.func @invokeLandingpad() -> i32 attributes { personality = @__gxx_personali
+  %14 = llvm.invoke @vararg_foo(%5, %5) to ^bb2 unwind ^bb1 vararg(!llvm.func<struct<(i32, f64, i32)> (i32, ...)>) : (i32, i32) -> !llvm.struct<(i32, f64, i32)>
+
+// CHECK: ^[[BB6:.*]]:
+// CHECK: %[[FUNCV:.*]] = llvm.mlir.addressof @vararg_foo : !llvm.ptr
```
nit: FUNCV -> FUNC?

gysit commented on mlir/test/Target/LLVMIR/llvmir.mlir:
```diff
@@ -1563,6 +1571,19 @@ llvm.func @invokeLandingpad() -> i32 attributes { personality = @__gxx_personali
+
+// CHECK: [[BB5:.*]]:
+// CHECK: invoke void (ptr, ...) undef(ptr %[[a1]], i32 0)
+// CHECK-NEXT: to label %[[normal:[0-9]+]] unwind label %[[unwind:[0-9]+]]
```
```suggestion
```
nit: i would probably drop that since it seems not relevant for this test?

gysit commented on mlir/test/Target/LLVMIR/llvmir.mlir:
```diff
@@ -1563,6 +1571,19 @@ llvm.func @invokeLandingpad() -> i32 attributes { personality = @__gxx_personali
+
+// CHECK: [[BB4:.*]]:
+// CHECK: invoke void (ptr, ...) @vararg_foo(ptr %[[a1]], i32 0)
+// CHECK-NEXT: to label %[[normal:[0-9]+]] unwind label %[[unwind:[0-9]+]]
```
```suggestion
```
nit: I would drop that since it seems not relevant for this test?

gysit reviewed (approved):
LGTM modulo nit comments!

Please make sure the CI is green before landing. There seems to be some formatting issue at the moment.

ivanradanov committed:
Apply suggestions from code review

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

ivanradanov committed:
FUNCV -> FUNC

ivanradanov committed:
clang-format

ivanradanov committed:
Add easy getter for the callee llvm function type in CallOp

Dinistro reviewed (approved):
The additional member function seems helpful. For consistency, I would also add it to invoke.

ivanradanov committed:
Add helper to invoke too and use it to simplify translation

Dinistro commented:
Thank you very much for addressing all our comments and for working on extending the LLVM dialect 😄
